sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -163,6 +163,155 @@ class CollationStringExpressionsSuite
     })
   }
 
+  test("Support Left/Right/Substr with implicit collation") {
+    case class SubstringTestCase(query: String, collation: String, result: Row)
+    val longString = "In the course of human events"
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        SubstringTestCase(s"select left(left('$longString' collate " + c + ", 5), 1)", c, Row("I")),
+        SubstringTestCase(
+          s"select right(right('$longString' collate " + c + ", 5), 1)", c, Row("s")),
+        SubstringTestCase(
+          s"select substr(substr('$longString' collate " + c + ", 4), 2)", c,
+          Row("he course of human events"))
+      )
+    )
+
+    checks.foreach { check =>
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    }
+  }
```

Comment:
I think this test is not needed

for the scope of this ticket (adding collation awareness to a string expression), it doesn't matter where the collationId came from (explicit or implicit), what matters is can the expression handle it properly

for example: if `left('In the course of human events' collate unicode, 5)` returns proper result value (`'In th'`) and type (`StringType(3)`), then calling left on that resulting string is no different than calling `left('In th' collate unicode, 1)`

Reply:
in other words, let's just look from the outside at string expression like a black box module, and focus just on making it work with collations properly

otherwise, we might end up with too many repetitive tests

Reply:
Will remove!